### PR TITLE
Enable reproducible archives in server and bootloader build

### PR DESCRIPTION
**Pull request by pmossman**
## What
I noticed that the `:buildDockerImage-server` and `:buildDockerImage-bootloader` tasks were never up-to-date, even with no changes between builds. [Discussion on Slack](https://airbytehq-team.slack.com/archives/C02TXQ020QM/p1646853607423209) led to a few fixes to enable these tasks to be up-to-date.

Addresses #11036 

## How
Definitely read the thread for detailed context, but in short, the airbyte-server and airbyte-bootloader subprojects use the shadowJar plugin, and that plugin apparently needs to be specifically configured to produce consistent archive outputs that can be recognized by gradle as unchanged.

Also, we were deleting the `build/docker` output directory before dockerCopy tasks, which caused some task cache misses. Now, we're using a Sync command to preserve unchanged files during this copy. 

Finally, it was necessary to add the reproducible archive task block to the subprojects that use shadowJar to actually get these tasks to cache. For some reason, other subprojects that _don't_ use the shadowJar plugin apparently don't need this, but there must be some interaction between shadowJar and other archive tasks that necessitates placing this block at the end of the subproject's build.gradle file.

This is the block I'm referring to when I say "reproducible archive task block":
```
tasks.withType(AbstractArchiveTask) {
    preserveFileTimestamps = false
    reproducibleFileOrder = true
}
```

In conclusion, with all of these changes, every `:buildDockerImage-*` task is up-to-date when re-building with no changes, which can _drastically_ speed up builds in some cases. I've seen 5-10 minutes wasted on these docker image build tasks that should be up-to-date, so hopefully this change improves our local developer experience quite a bit!

## Recommended reading order
Top to bottom is fine

### CODE REVIEW

**Review by jdpgrailsdev — APPROVED** [review]
:shipit:

File: airbyte-bootloader/build.gradle [hunk 1]
```diff
@@ -81,3 +83,10 @@ task copyGeneratedTar(type: Copy) {
 Task dockerBuildTask = getDockerBuildTask("bootloader",  "$project.projectDir")
 dockerBuildTask.dependsOn(copyGeneratedTar)
 assemble.dependsOn(dockerBuildTask)
+
+// produce reproducible archives
+// (see https://docs.gradle.org/current/userguide/working_with_files.html#sec:reproducible_archives)
+tasks.withType(AbstractArchiveTask) {
```
Comment (jrhizor) [inline airbyte-bootloader/build.gradle:89]: Are these changes and the ones on 32-33 required or just the ones on 32-33?

File: airbyte-server/build.gradle [hunk 2]
```diff
@@ -141,6 +145,13 @@ task copyGeneratedTar(type: Copy) {
...
 }
 
-Task dockerBuildTask = getDockerBuildTask("server",  "$project.projectDir")
+Task dockerBuildTask = getDockerBuildTask("server", "$project.projectDir")
 dockerBuildTask.dependsOn(copyGeneratedTar)
 assemble.dependsOn(dockerBuildTask)
+
+// produce reproducible archives
+// (see https://docs.gradle.org/current/userguide/working_with_files.html#sec:reproducible_archives)
+tasks.withType(AbstractArchiveTask) {
```
Comment (jrhizor) [inline airbyte-server/build.gradle:154]: same as above

**Review by jrhizor — APPROVED** [review]

**Comment by jrhizor** [comment]
Should the "reproducible archive task block" be part of other build subprojects besides bootloader/server?

**Comment by pmossman** [comment]
> Should the "reproducible archive task block" be part of other build subprojects besides bootloader/server?

Yeah @jrhizor I'm confused why this doesn't seem to be needed in other subprojects. My hunch is that it's only needed in the subprojects that use shadowJar, but I don't have a great justification for why that would be. It definitely wouldn't hurt to add them to all subprojects that publish jars, so I can go through and add that block to be safe. What do you think?

**Comment by jrhizor** [comment]
Oh yeah, forgot that both of these had shadowjars. 🤦 

I think this is fine as is. When we end up improving how we publish jars I'm sure this will come up.

**Comment by davinchia** [comment]
agree with Jared this is fine as is since we'll should get rid of this in the coming months.

@pmossman some context here - the shadow plugin packages all dependencies into one jar vs the usual and better practice of leaving the dependencies in their own jars and placing the various jars in a location that can be found (the usual Gradle application plugin). Because of this, Gradle cannot rely on the outputs of the other modules it depends on and checks according to file order + timestamp. great work catching this!
